### ISSUE TITLE
Bridges: enable chopsticks to emulate passing XCM messages over bridge

### ISSUE THREAD

**Issue by serban300** [paritytech/polkadot-sdk#4793]
In order to send messages over the bridge we need to also pass a finality proof and a storage proof. Chopsticks doesn't support this at the moment and it would be quite complex to implement.

One idea would be to expose a runtime API for dispatching bridged messages and then chopsticks will be able to call this runtime API, get storage diffs, and actually build a block with those storage diffs enacted

We can explore other options as well

Related chopsticks issue:  https://github.com/AcalaNetwork/chopsticks/issues/741

**Comment by xlc** [comment]
Another alternative will be having a root origin call to apply an arbitrary message without proof. This could also be used by governance for whatever reason. And then we should be able to fake a DMP from relaychain to dispatch call from root origin to dispatch messages.

**Comment by bkontur** [comment]
just thinking out loud, yeah that `dry_run` messages runtime API is not a bad idea,
[here](https://github.com/AcalaNetwork/chopsticks/issues/741#issuecomment-2165017007) I wrote how should be possible to read message (`BridgeMessage(dest, xcm) as Vec<u8>`) on source BH,

and then we could expose [`MessageDispatch`](https://github.com/paritytech/polkadot-sdk/blob/master/cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_westend_config.rs#L219) as this runtime API on target BH, because the real message dispatch should hit this code: https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/universal_exports.rs#L428-L434, which decodes `BridgeMessage(dest, xcm) as Vec<u8>` and sends it with `XcmRouter`, which is using `XcmpQueue`, so the `XcmpQueue` makes storage diff which we need.

**Comment by bkontur** [comment]
> Another alternative will be having a root origin call to apply an arbitrary message without proof. This could also be used by governance for whatever reason. And then we should be able to fake a DMP from relaychain to dispatch call from root origin to dispatch messages.

@xlc I think if we allow `pallet_xcm::Config::SendXcmOrigin` for a governance on target BH, it should be "almost" enough. We can decode `BridgeMessage(dest, xcm) as Vec<u8>` from source BH, but we need to prepend it with https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/universal_exports.rs#L428-L434 to be valid on the target AssetHub

**Comment by bkontur** [comment]
There is ongoing work for XCM dry run: https://github.com/paritytech/polkadot-sdk/pull/6002

**Comment by raymondkfcheung** [comment]
Maybe related to #6572

**Comment by raymondkfcheung** [comment]
# Summary of Work and Next Steps

I've been exploring [**testing XCM over Bridges**](https://github.com/raymondkfcheung/xcm-over-bridges) using Chopsticks and have documented the process and findings here:
* Documents how to **spin up Polkadot + Kusama local stacks** (Asset Hubs & Bridge Hubs) via Chopsticks
* Shows **manual XCM decoding**, step-by-step [**transaction walkthroughs**](https://github.com/raymondkfcheung/xcm-over-bridges/blob/main/walkthrough.md), and **Vitest specs** for ["transfers across Bridges"](https://github.com/raymondkfcheung/xcm-over-bridges/blob/main/tests/bridges/xcm-over-bridges.test.ts#L157)
* Explains [**why end-to-end testing is hard right now**](https://github.com/raymondkfcheung/xcm-over-bridges/blob/main/testing.md), since `BridgePolkadotMessages.receive_messages_proof` requires a valid proof that's currently only generated by live relayers.

## Next Steps / Options

To test this without manual decoding or real relayers, we'll need a way to simulate proofs:

1. Introduce a **test-only extrinsic** (e.g. `receive_messages_proof_mock`) to inject a mock proof into `BridgePolkadotMessages` during local tests.
2. Add mock-proof capability **inside Chopsticks**, so it can fake the bridge proof handling when simulating message delivery. Relevant follow-ups:
   * https://github.com/AcalaNetwork/chopsticks/issues/681
   * https://github.com/AcalaNetwork/chopsticks/issues/741

**Comment by acatangiu** [comment]
> Introduce a test-only extrinsic (e.g. receive_messages_proof_mock) to inject a mock proof into BridgePolkadotMessages during local tests.

Yes, this is the easiest approach. You can add root-guarded extrinsics: `force_submit_messages_proof()` which can effectively be called only by governance or in testing.

> Add mock-proof capability inside Chopsticks, so it can fake the bridge proof handling when simulating message delivery. 

I think this is overly complicated and too bridge specific for Chopsticks

-----

The current https://github.com/raymondkfcheung/xcm-over-bridges is too complicated because it requires running relayers to pass along proofs. With option 1 above, chopsticks can simply inject the proofs itself and simplify everything.
